pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -245,9 +306,149 @@ func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
 		return err
 	}
 
+	if err := r.invalidateLease(nodeName); err != nil {
+		return err
+    }
+
 	return nil
 }
 
+func (r *ReconcileNodeMaintenance) invalidateLease(nodeName string) error {
+		if r.isLeaseSupported {
+				log.Info("Lease object supported,  invalidating lease")
+
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+				lease := &coordv1beta1.Lease{}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return err
+				}
+				lease.Spec.AcquireTime = nil
+				lease.Spec.LeaseDurationSeconds = nil;
+				lease.Spec.RenewTime = nil
+				lease.Spec.LeaseTransitions = nil
+
+				if err := r.client.Update(context.TODO(), lease); err != nil {
+					return err
+				}
+		}
+		return nil
+}
+
+func (r *ReconcileNodeMaintenance) obtainLease(node *corev1.Node, durationInSeconds int) (reconcile.Result, error) {
+
+	lease, needUpdate, err := r.createOrGetExistingLease(node, durationInSeconds)
+
+	if err != nil {
+		log.Errorf("MaintOn: failed to create or get existing lease error=%v", err)
+		return reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}, err
+	}
+
+	if needUpdate {
+
+		log.Info("MaintOn: update lease - lease duration change requested")
+
+		if _, err = r.updateLease(node, lease, durationInSeconds, TransitionNoChange, false); err != nil {
+		return reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNodeMaintenance) createOrGetExistingLease(node *corev1.Node, durationInSeconds int) (*coordv1beta1.Lease, bool, error) {
+	nodeName := node.ObjectMeta.Name
+
+	holderIdentity := LeaseHolderIdentity
+
+	owner := &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        node.ObjectMeta.UID,
+	}
+	leaseTransitions := int32(1)
+	microTimeNow := metav1.NewMicroTime(time.Now())
+
+	tmpDurationInSeconds := int32(durationInSeconds)
+	var refDuration *int32 =  &tmpDurationInSeconds
+
+	lease := &coordv1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            nodeName,
+			Namespace:       LeaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1beta1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: refDuration,
+			AcquireTime:          &microTimeNow,
+			RenewTime:            &microTimeNow,
+			LeaseTransitions:     &leaseTransitions,
+		},
+	}
+
+	if err := r.client.Create(context.TODO(), lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return  nil, false, err
+				}
+				return lease, true, nil
```

Comment:
Wouldn't it be better to try `Get()` before constructing a new one?

Reply:
I think checking and then creating the lease if it doesn't already exist is more atomic - there is no period between the check and creation where another party could get it. In any event I had it like that before (first checking if it doesn't exist and then create the lease), but you didn't like the flow, so I changed it per your request.

Reply:
I was mostly talking about the higher-level flow of having a function that returned a lease whether one already existed or not.